### PR TITLE
New component: @reach/checkbox

### PR DESCRIPTION
**Pull request by chancestrickland**
Thank you for contributing to Reach UI! Please fill in this template before submitting your PR to help us process your request more quickly.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code (Compile and run).
- [ ] Add or edit tests to reflect the change (Run with `yarn test`).
- [x] Add or edit Storybook examples to reflect the change (Run with `yarn start`).
- [x] Ensure formatting is consistent with the project's Prettier configuration.

This pull request:

- [x] Creates a new package
- [ ] Fixes a bug in an existing package
- [ ] Adds additional features/functionality to an existing package
- [ ] Updates documentation or example code
- [ ] Other

If creating a new package:

- [x] Make sure the new package directory contains each of the following, and that their structure/formatting mirrors other related examples in the project:
  - [x] `examples` directory
  - [x] `src` directory with an `index.js` entry file
  - [ ] Type definitions in an `index.d.ts` file are desired but not required for the PR to be merged **TODO**
  - [x] At least one example file per feature introduced by the new package
  - [x] Base styles in a `style.css` file (if needed by the new package)

### CODE REVIEW

File: packages/checkbox/examples/group.example.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import React from "react";
```
Comment (ryanflorence) [inline packages/checkbox/examples/group.example.js]: What does this illustrate other than that you can have multiple checkboxes?

File: packages/checkbox/src/index.js [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+  );
+});
+
+if (__DEV__) {
+  CustomCheckbox.propTypes = {
+    checkmarks: shape({
+      true: node,
+      false: node,
+      mixed: node
+    })
```
Comment (ryanflorence) [inline packages/checkbox/src/index.js]: shouldn't this be required?

File: packages/checkbox/src/index.js [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+      readOnly={readOnly}
+      type="checkbox"
+    />
+  );
+});
+
+if (__DEV__) {
+  MixedCheckbox.propTypes = {
+    checked: oneOfType([bool, string]),
+    defaultChecked: bool,
```
Comment (ryanflorence) [inline packages/checkbox/src/index.js]: should this be `oneOfType([bool, string])`?

File: packages/checkbox/src/index.js [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    disabled: bool,
```
Comment (ryanflorence) [inline packages/checkbox/src/index.js]: Do we need the rest of these prop-types? They're all just normal checkbox proptypes (and React, afaik) doesn't give us a way to extend "checkbox proptypes".

Additionally I don't think name or value need to be required. They aren't required in normal checkboxes, what makes this different?

File: packages/checkbox/src/index.js (hunk 2 above)
Comment (chancestrickland) [reply packages/checkbox/src/index.js]: I don't think so. You could style a custom checkbox without passing any additional components (with pseudo elements, for example).

**Comment by ryanflorence** [comment]
I made an example of an uncontrolled custom checkbox and I'm getting this error:

```
MixedCheckbox contains an input of type checkbox with both checked and defaultChecked props. Input elements must be either controlled or uncontrolled
```

```jsx
     <CustomCheckbox
        defaultChecked={state}
        onChange={event => {
          const { checked } = event.target;
          setState(checked);
        }}
        checkmarks={{
          true: <span style={smileyStyle}>😃</span>,
          false: <span style={smileyStyle}>🙁</span>
        }}
      />
```

While we can fix this directly, I think the issue stems from a deeper composition problem that might reveal more problems down the line.

`MixedCheckbox` is entangled with `CustomCheckbox`. I think the code could be organized in a way that `CustomCheckbox` looks at its props and then decides to render a normal checkbox or a mixed checkbox, rather than mixed checkbox being aware of custom checkboxes.

`MixedCheckbox` can have a much simpler implementation because it:

1. won't have any knowledge of custom checkboxes
2. won't need to support being uncontrolled, all mixed checkboxes are necessarily controlled.

The current code has a cartesian product of requirements that must be dealt with, rather than things being encapsulated and then composed.

File: packages/checkbox/src/index.js [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+
+////////////////////////////////////////////////////////////////////////////////
+export const CustomCheckbox = forwardRef(function CustomCheckbox(
+  {
+    checked: controlledChecked,
+    defaultChecked,
+    disabled: disabledProp,
+    checkmarks,
+    children,
+    id,
```
Comment (ryanflorence) [inline packages/checkbox/src/index.js]: any reason we're messing around with `id`, `name` here? They just get forwarded to the underlying input.

Comment (chancestrickland) [reply packages/checkbox/src/index.js]: The rest of the props are forwarded to the wrapper, but even for a high-level `CustomCheckbox` abstraction users will regularly need to pass `id`, `name` and `value` to the input

File: packages/checkbox/examples/group.example.js (hunk 1 above)
Comment (chancestrickland) [reply packages/checkbox/examples/group.example.js]: Not much without the `CheckboxGroup` component! I removed this one.

File: packages/checkbox/src/index.js (hunk 3 above)
Comment (chancestrickland) [reply packages/checkbox/src/index.js]: Fixed

**Comment by chancestrickland** [comment]
@ryanflorence I updated the MixedCheckbox to remove any reference to CustomCheckbox completely, simplifying it to its singular purpose as we discussed. Take another look when you have a moment!

**Comment by ryanflorence** [comment]
If you click the readOnly checkbox twice it switches over to checked

**Pull request merged**